### PR TITLE
Jin

### PR DESCRIPTION
**Pull request by Jin4202**
Use exceptions to handle error now

### CODE REVIEW

File: app/src/main/java/com/example/calculator/Calculation.java [hunk 1]
```diff
@@ -7,14 +7,45 @@
...
+        if(opening != closing) {
+            int addParenthesis = opening - closing;
+            int repeat = Math.abs(addParenthesis);
+            for(int i = 0; i < repeat; i++) {
+                if(addParenthesis > 0) {
+                    mStringArrayList.add(new Parenthesis(')'));
+                } else {
+                    mStringArrayList.add(0, new Parenthesis('('));
+                }
+            }
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/Calculation.java]: Could you change this more effeciently?

File: app/src/main/java/com/example/calculator/Calculation.java [hunk 2]
```diff
@@ -7,14 +7,45 @@
...
+        }
+    }
+
+    public double solveAnswer() throws InvalidFormatError{
         ArrayList<Token> tokenArr = mStringArrayList;
+        ArrayList<Token> answer = solveAnswer(tokenArr);
+        if(answer.size() == 0) {
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/Calculation.java]: You can use isEmpty().

File: app/src/main/java/com/example/calculator/Calculation.java [hunk 3]
```diff
@@ -46,15 +78,19 @@ private boolean isParenthesisExist(ArrayList<Token> tokenArr) {
...
-
-    private void solve(ArrayList<Token> tokenArr, Class<?> c) {
+    private void solve(ArrayList<Token> tokenArr, Class<?> c) throws InvalidFormatError{
         for(int i = 0; i < tokenArr.size(); i++) {
             Token current = tokenArr.get(i);
             if(c.isInstance(current)) {
                 OperatorToken operatorTokenToken = (OperatorToken) current;
-                NumberToken solvedToken = operatorTokenToken.calculate(tokenArr.get(i-1), tokenArr.get(i+1));
 
+                if(i+1 >= tokenArr.size() || i == 0 || tokenArr.get(i+1) instanceof OperatorToken) {
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/Calculation.java]: Could you comment what each condition means?

File: app/src/main/java/com/example/calculator/Calculation.java [hunk 4]
```diff
@@ -111,15 +146,15 @@ private void solve(ArrayList<Token> tokenArr, Class<?> c) {
     }
 
     private ArrayList<Token> subArrayList(ArrayList<Token> arr, int fromIndex, int toIndex) {
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/Calculation.java]: You can use ArrayList.subList()

File: app/src/main/java/com/example/calculator/Divide.java [hunk 5]
```diff
@@ -8,7 +10,7 @@ public Divide() {
     public NumberToken calculate(Token num1, Token num2) {
         double num2safe = num2.getNumber();
         if(Double.compare(num2safe,0) == 0) {
-            num2safe = 1;
+            throw new ArithmeticException();
         }
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/Divide.java:14]: You don't need this by yourselves.
When you divide by zero, JVM will throw the exception.

File: app/src/main/java/com/example/calculator/MainActivity.java [hunk 6]
```diff
@@ -173,9 +172,17 @@ public void onClick(View view) {
         buttonSolve.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View view) {
+                if(inputText.length() == 0) {
```
Comment (solderman0) [inline app/src/main/java/com/example/calculator/MainActivity.java:180]: isEmpty()?

**Review by solderman0 — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/calculator/Divide.java (hunk 5 above)
Comment (Jin4202) [reply app/src/main/java/com/example/calculator/Divide.java:14]: https://stackoverflow.com/questions/14137989/java-division-by-zero-doesnt-throw-an-arithmeticexception-why

**Review by solderman0 — APPROVED** [review]

**Pull request merged**